src/extension.ts

```diff
@@ -26,60 +27,84 @@ export function activate(context: vscode.ExtensionContext) {
         }
-    }));
-    context.subscriptions.push(vscode.commands.registerCommand('appService.Restart', (node: AppServiceNode) => {
+    });
+    initAsyncCommand(context, 'appService.Restart', async (node: AppServiceNode) => {
         if (node) {
             outputChannel.appendLine(`Restarting App "${node.site.name}"...`);
-            node.restart(azureAccount).then(() => outputChannel.appendLine(`App "${node.site.name}" has been restarted.`), err => outputChannel.appendLine(err));
+            await node.restart(azureAccount).then(() => outputChannel.appendLine(`App "${node.site.name}" has been restarted.`), err => outputChannel.appendLine(err));
         }
-    }));
-    context.subscriptions.push(vscode.commands.registerCommand('appService.CreateWebApp', async () => {
+    });
+    initAsyncCommand(context, 'appService.CreateWebApp', async () => {
         const wizard = new WebAppCreator(outputChannel, azureAccount);
         const result = await wizard.run();
         
         if (result.status === 'Completed') {
             vscode.commands.executeCommand('appService.Refresh');
         }
-    }));
-    context.subscriptions.push(vscode.commands.registerCommand('appService.DeployZipPackage', async (context: any) => {
+    });
+    initAsyncCommand(context,'appService.DeployZipPackage', async (context: any) => {
         if (context instanceof AppServiceNode) {
             const wizard = new WebAppZipPublisher(outputChannel, azureAccount, context.subscription, context.site);
             await wizard.run();
         } else if (context instanceof vscode.Uri) {
             const wizard = new WebAppZipPublisher(outputChannel, azureAccount, null, null, context.fsPath, null);
             await wizard.run();
         }
-    }));
-    context.subscriptions.push(vscode.commands.registerCommand('appService.ZipAndDeploy', async (context: any) => {
+    });
+    initAsyncCommand(context, 'appService.ZipAndDeploy', async (context: any) => {
         if (context instanceof vscode.Uri) {
             const folderPath = context.fsPath;
             const wizard = new WebAppZipPublisher(outputChannel, azureAccount, null, null, null, folderPath);
             await wizard.run();
         }
-    }));
+    });
 }
 
 export function deactivate() {
 }
+
+function initCommand(context: vscode.ExtensionContext, commandId: string, callback: (...args: any[]) => any) {
+    initAsyncCommand(context, commandId, (...args: any[]) => Promise.resolve(callback(...args)));
+}
+
+function initAsyncCommand(context: vscode.ExtensionContext, commandId: string, callback: (...args: any[]) => Promise<any>) {
+    context.subscriptions.push(vscode.commands.registerCommand(commandId, async (...args: any[]) => {
+        const start = os.uptime();
+        let result = 'Succeeded';
+        let errorData: string = '';
+
+        try {
+            await callback(...args);
+        } catch (err) {
+            result = 'Failed';
+            errorData = JSON.stringify(err);
```

Comment:
A few things on this:
1. When I run `JSON.stringify(err)` I get back `{}`. Are we sure this is a JSON object that can be stringified?
1. Even if we could stringify all of `err`, I think we should only track `err.message`. For example, the stack trace looks like this and you can clearly see my username in the path (I'm worried that's PII we don't want)
```
Error: fdsafdsa
	at AppInsightsAsyncCorrelatedErrorWrapper (/Users/ericjizba/.vscode/extensions/msazurermtools.azurerm-vscode-tools-0.3.6/node_modules/applicationinsights/out/AutoCollection/CorrelationContextManager.js:172:18)
	at /Users/ericjizba/.vscode/extensions/ms-azuretools.vscode-cosmosdb-0.0.1/out/src/extension.js:147:15
	at Generator.next (<anonymous>)
	at /Users/ericjizba/.vscode/extensions/ms-azuretools.vscode-cosmosdb-0.0.1/out/src/extension.js:11:71
	at new ZoneAwarePromise (/Users/ericjizba/.vscode/extensions/msazurermtools.azurerm-vscode-tools-0.3.6/node_modules/zone.js/dist/zone-node.js:551:29)
	at __awaiter (/Users/ericjizba/.vscode/extensions/ms-azuretools.vscode-cosmosdb-0.0.1/out/src/extension.js:7:12)
	at copyConnectionString (/Users/ericjizba/.vscode/extensions/ms-azuretools.vscode-cosmosdb-0.0.1/out/src/extension.js:146:12)
	at initAsyncCommand (/Users/ericjizba/.vscode/extensions/ms-azuretools.vscode-cosmosdb-0.0.1/out/src/extension.js:50:74)
	at /Users/ericjizba/.vscode/extensions/ms-azuretools.vscode-cosmosdb-0.0.1/out/src/extension.js:74:19
	at Generator.next (<anonymous>)
	at /Users/ericjizba/.vscode/extensions/ms-azuretools.vscode-cosmosdb-0.0.1/out/src/extension.js:11:71
	at new ZoneAwarePromise (/Users/ericjizba/.vscode/extensions/msazurermtools.azurerm-vscode-tools-0.3.6/node_modules/zone.js/dist/zone-node.js:551:29)
	at __awaiter (/Users/ericjizba/.vscode/extensions/ms-azuretools.vscode-cosmosdb-0.0.1/out/src/extension.js:7:12)
	at context.subscriptions.push.vscode.commands.registerCommand (/Users/ericjizba/.vscode/extensions/ms-azuretools.vscode-cosmosdb-0.0.1/out/src/extension.js:69:85)
```

Reply:
You are right, JSON.stringify is not the solution. Added a helper method for this, some sample outputs:
```
util.errToString(error)
"{"Error":"UserCancelledError","Message":""}"
util.errToString('evil string')
"evil string"
util.errToString(123.456)
"123.456"
util.errToString(true)
"true"
util.errToString(false)
"false"
util.errToString(Date.today())
"{"object":"Date","toString":"Thu Sep 14 2017 00:00:00 GMT-0700 (Pacific Daylight Time)"}"
util.errToString([1, 2, 3, 4])
"[1,2,3,4]"
util.errToString(/^A-Z/g)
"{"object":"RegExp","toString":"/^A-Z/g"}"
```